_posts/2020-07-31-fnn-vae-for-noisy-timeseries/fnn_vae_for_noisy_timeseries.Rmd

```diff
@@ -0,0 +1,509 @@
+# add noise
+noise <- 1 # also used 1.5, 2, 2.5
+roessler <- roessler + rnorm(10000, mean = 0, sd = noise)
+
+```
+
+Here you can compare effects of adding no (left), standard deviation-1 (middle), and standard deviation-2.5 Gaussian noise:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Roessler series with added noise. Top: none. Middle: SD = 1. Bottom: SD = 2.5."}
+knitr::include_graphics("images/roessler_noise.png")
+```
+
+Otherwise, preprocessing proceeds as in the previous posts. In the upcoming results section, we'll compare forecasts not just
+to the "real", after noise addition, test split of the data, but also to the underlying Roessler system -- that is, the thing
+we're really interested in. (Just that in the real world, we can't do that check.) This second test set is prepared for
+forecasting just like the other one; to avoid duplication we don't reproduce the code.
+
+```{r, eval = FALSE}
+n_timesteps <- 120
+batch_size <- 32
+
+gen_timesteps <- function(x, n_timesteps) {
+  do.call(rbind,
+          purrr::map(seq_along(x),
+                     function(i) {
+                       start <- i
+                       end <- i + n_timesteps - 1
+                       out <- x[start:end]
+                       out
+                     })
+  ) %>%
+    na.omit()
+}
+
+train <- gen_timesteps(roessler[1:(n/2)], 2 * n_timesteps)
+test <- gen_timesteps(roessler[(n/2):n], 2 * n_timesteps) 
+
+dim(train) <- c(dim(train), 1)
+dim(test) <- c(dim(test), 1)
+
+x_train <- train[ , 1:n_timesteps, , drop = FALSE]
+y_train <- train[ , (n_timesteps + 1):(2*n_timesteps), , drop = FALSE]
+
+ds_train <- tensor_slices_dataset(list(x_train, y_train)) %>%
+  dataset_shuffle(nrow(x_train)) %>%
+  dataset_batch(batch_size)
+
+x_test <- test[ , 1:n_timesteps, , drop = FALSE]
+y_test <- test[ , (n_timesteps + 1):(2*n_timesteps), , drop = FALSE]
+
+ds_test <- tensor_slices_dataset(list(x_test, y_test)) %>%
+  dataset_batch(nrow(x_test))
+```
+
+## Results
+
+The LSTM used for comparison with the VAE described above is identical to the architecture employed in the previous post.
+While with the VAE, an `fnn_multiplier` of 1 yielded sufficient regularization for all noise levels, some more experimentation
+was needed for the LSTM: At noise levels 2 and 2.5, that multiplier was set to 5.
+
+As a result, in all cases, there was one latent variable with high variance and a second one of minor importance. For all
+others, variance was close to 0.
+
+*In all cases* here means: In all cases where FNN regularization was used. As already hinted at in the introduction, the main
+regularizing factor providing robustness to noise here seems to be FNN loss, not KL divergence. So for all noise levels,
+besides FNN-regularized LSTM and VAE models we also tested their non-constrained counterparts.
+
+#### Low noise
+
+Seeing how all models performed bravourously on the original deterministic series, a noise level of 1 can almost be treated as
```

Comment:
I don't know the word bravourously. 😳

Reply:
oops - that happens when you think "yeah that must exist in English as well, as it's from French anyway" and are too lazy to look it up :-(
guess i need sth like "superbly"